Support Attributes for FinalClassFixer

andythorne opened the pull request:
> This is a rebase of https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6622, which seems to work great. If it's missing something I'm happy to add it.

## Description
This PR brings annotation support and a number of refinements.

Replaces: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6622
Close: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/5782

- annotation_include has been deprecated and renamed to include
- annotation_exclude has been deprecated and renamed to exclude

If you provide:

- default config -> runs with annotation support
- old config and new config, both not default config, throws a config-exception
- old config that is not default config -> runs without annotation support
- new config that is not default config -> runs with annotation support

If you provide `include` than if any annotion or the PHPDoc of the class contains at least one as configured, but not necessary all, it will be fixed.
If you provide `exclude` than if any annotion or the PHPDoc of the class contains at least one as configured, but not necessary all, it will not be fixed.
```

andythorne commented:
This PR should now have all the tests and phpstan issues resolved now

Wirone commented:
@andythorne please make CI green so we can progress to code review 🙂.

andythorne commented:
Thanks, @Wirone, I didn't realise deprecations fail the tests! it's all green now on my local with my latest commit 🙏

Wirone commented on tests/Fixer/ClassNotation/FinalClassFixerTest.php:
```diff
@@ -37,13 +37,20 @@ public static function provideFixCases(): array
         return [
             ['<?php /** @Entity */ class MyEntity {}'],
+            ['<?php #[Entity] class MyEntity {}'],
             ['<?php use Doctrine\ORM\Mapping as ORM; /** @ORM\Entity */ class MyEntity {}'],
+            ['<?php use Doctrine\ORM\Mapping as ORM; #[ORM\Entity] class MyEntity {}'],
             ['<?php use Doctrine\ORM\Mapping; /** @Mapping\Entity */ class MyEntity {}'],
             ['<?php use Doctrine\ORM; /** @ORM\Mapping\Entity */ class MyEntity {}'],
+            ['<?php use Doctrine\ORM; #[ORM\Mapping\Entity] class MyEntity {}'],
             ['<?php /** @Document */ class MyDocument {}'],
+            ['<?php #[Document] class MyDocument {}'],
             ['<?php use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM; /** @ODM\Document */ class MyEntity {}'],
+            ['<?php use Doctrine\ODM\MongoDB\Mapping\Annotations as ODM; #[ODM\Document] class MyEntity {}'],
             ['<?php /** @entity */ class MyEntity {}'],
-            ['<?php use Doctrine\ORM\Mapping as ORM; /** @orm\entity */ class MyEntity {}'],
+            ['<?php #[Entity] class MyEntity {}'],
+            ['<?php use Doctrine\ORM\Mapping as ORM; /** @ORM\entity */ class MyEntity {}'],
+            ['<?php use Doctrine\ORM\Mapping as ORM; #[ORM\entity] class MyEntity {}'],
```
Cases with attributes must be moved to separate test/provider and run only on PHP >= 8.0 - on 7.4 it is treated as comment, so effectively the code is like `<?php use Doctrine\ORM\Mapping as ORM;` and I believe such test does not provide value.

Just look for `testFix80` and `provideFix80Cases` in the project to see how it's done.

Wirone reviewed (changes requested):
Did not review everything, but I want to provide fast feedback with things that need to be improved 🙂.

andythorne commented:
@Wirone thanks for the feedback! I've implemented the changes you've requested 🙏

Wirone reviewed (changes requested):
Docs need polishing, reusing defaults is a cosmetic change (nice to have, if it makes sense). I hope it's not a problem for you that review is in rounds, I still need to focus on actual changes in the fixer itself. But I do it in the meantime 😅.

andythorne commented:
Thank you, I greatly appreciate any time you can give to review, it's never a problem for me but sorry if it's breaking up your flow! I'll get these updated, they all make sense to me - I'll have to pick it up after the weekend though, should have some changes on Monday all being well.

Wirone commented:
You know, some people prefer to get all comments at once and fix them in one session, instead of coming back multiple times. But it's sometimes hard to review everything, especially because I do it for example while walking with dog (like now) 😆. So I'm happy it's not a problem for you 👍.

andythorne commented:
I saw the original PR had changed some of the docs, so I reverted back to master's branch and just updated them to reference annotations _and_ attributes instead, so hopefully it is more readable now!

Also implemented the defaults from the constants.

Wirone commented on src/Fixer/ClassNotation/FinalInternalClassFixer.php:
```diff
@@ -195,23 +237,41 @@ private function isClassCandidate(TokensAnalyzer $tokensAnalyzer, Tokens $tokens
+        $index = $tokens->getPrevNonWhitespace($index);
+
+        if ($this->checkAttributes && $tokens[$index]->isGivenKind(CT::T_ATTRIBUTE_CLOSE)) {
+            if (!$this->isClassCandidateBasedOnAttribute($tokens, $index)) {
+                return false;
+            }
+
+            while ($tokens[$index]->isGivenKind(CT::T_ATTRIBUTE_CLOSE)) {
```
Is it possible to refactor it in a way, where `while ($tokens[$attributeCloseIndex]->isGivenKind(CT::T_ATTRIBUTE_CLOSE))` is looped only once? Because `isClassCandidateBasedOnAttribute()` internally also does the same `while`.

Maybe this method should return `?int` instead, and `return false;` should be done on returned `null`, and `int` should be used as `$index`? Thinking out loud 🙂. I'm just wondering if we need 2 similar loops.

andythorne replied:
I agree it wasn't easy to read. Looking at the code we don't actually need the index in the return, so I've refactored the double-loop out.

Wirone commented on tests/Fixer/ClassNotation/FinalClassFixerTest.php:
```diff
@@ -75,4 +75,71 @@ public static function provideFixCases(): array
+        $this->doTest($expected, $input);
+    }
+
+    public static function provideFix82Cases(): array
```
Can we add some "dirty" cases, not only optimistic ones? Like with comments somewhere in-between etc.

```php
[
    '<?php use Doctrine\ORM; #[ORM\Mapping\Entity] readonly /* ... */ final class MyEntity {}',
    '<?php use Doctrine\ORM; #[ORM\Mapping\Entity] readonly /* ... */ class MyEntity {}',
],
```

Also multiline, maybe comments between attributes and class itself?

Wirone commented on tests/Fixer/ClassNotation/FinalClassFixerTest.php:
If I see correctly, there are no tests for mixed annotations+attributes usage. Is it possible at all? Like:

```php
/**
 * @A
 */
#[B]
class Foo {}
```

andythorne commented:
@Wirone Apologies for the delay, I've been busy! I've updated the PR with more test cases and a refactor of the candidate methods

Wirone commented on tests/Fixer/ClassNotation/FinalInternalClassFixerTest.php:
```diff
@@ -506,6 +506,18 @@ class Foo {}',
             ],
         ];
+
+        yield 'positive on doc, but attribute on doc' => [
```
```suggestion
        yield 'include by phpDoc, but exclude by attribute' => [
```

Wirone commented on tests/Fixer/ClassNotation/FinalInternalClassFixerTest.php:
```diff
@@ -526,10 +538,16 @@ public function testFix82(string $expected, ?string $input, array $config): void
     public static function provideFix82Cases(): iterable
     {
-        yield [
+        yield 'readonly attribute' => [
```
Maybe too long, but it's hard to name this case properly. Anyway there's no "attribute" here so it's wrong.
```suggestion
        yield 'readonly with enabled `consider_absent_docblock_as_internal_class`' => [
```

Wirone commented on tests/Fixer/ClassNotation/FinalInternalClassFixerTest.php:
```diff
@@ -526,10 +538,16 @@ public function testFix82(string $expected, ?string $input, array $config): void
             ['consider_absent_docblock_as_internal_class' => true],
         ];
+
+        yield 'readonly attribute with comment' => [
```
```suggestion
        yield 'readonly with `internal` attribute and comment in-between' => [
```

Wirone reviewed (changes requested):
I still need to review updated fixer's logic, but was able to take a quick look on tests, so here are small changes to make. Also, please rebase branch on latest `master` (and ignore fail for PHP 8.3).

Wirone commented on src/Fixer/ClassNotation/FinalInternalClassFixer.php:
```diff
@@ -177,23 +219,39 @@ private function isClassCandidate(TokensAnalyzer $tokensAnalyzer, Tokens $tokens
+        }
+
+        if (false === $attributeDecision || false === $phpDocDecision) {
+            return false;
+        }
+
+        if (true === $attributeDecision || true === $phpDocDecision) {
+            return true;
         }
 
-        $doc = new DocBlock($docToken->getContent());
+        return $this->configuration['consider_absent_docblock_as_internal_class'];
```
If I'm not mistaken, there's no possibility to reach this `return` - if both `$attributeDecision` and `$phpDocDecision` are `false`, `isClassCandidate()` will return `false`; if one of them is `true`, `isClassCandidate()` will return `false` anyway, if both are `true` then `isClassCandidate()` will return `true`. There's no path for going further than both conditions above 🤔.

andythorne replied:
Both `$attributeDecision` and `$phpDocDecision` can be null (their initial value) if no decision was given, which will trigger this case

Wirone replied:
Ah OK, I looked at those methods' return types and they're `bool`, but you're right variables are overridden only in specific circumstances so they may be `null` 👍.

Wirone commented on src/Fixer/ClassNotation/FinalInternalClassFixer.php:
```diff
@@ -177,23 +219,39 @@ private function isClassCandidate(TokensAnalyzer $tokensAnalyzer, Tokens $tokens
 
-        if (!$docToken->isGivenKind(T_DOC_COMMENT)) {
-            return $this->configuration['consider_absent_docblock_as_internal_class'];
+        if ($this->checkAttributes && $attributeIndex = $tokens->getPrevTokenOfKind($index, [[CT::T_ATTRIBUTE_CLOSE]])) {
```
Please add test case where there are 2+ classes in one file and first has attribute while second doesn't. I believe rewinding to first `CT::T_ATTRIBUTE_CLOSE` may lead to context switch, while we need to stick to class that triggered the fixer. It would be good to support such scenario, but I'm not sure if other fixers handle that properly, so if not, we can skip this.

There is already test with multiple classes, but each has attribute, so it works properly because it's optimistic scenario 😉.

Wirone commented on src/Fixer/ClassNotation/FinalInternalClassFixer.php:
```diff
@@ -202,12 +260,93 @@ private function isClassCandidate(TokensAnalyzer $tokensAnalyzer, Tokens $tokens
         }
 
-        return true;
+        foreach ($attributeCandidates as $attributeCandidate => $true) {
```
Maybe like this?

```suggestion
        foreach (array_keys($attributeCandidates) as $attributeCandidate) {
```

There's no point in defining `$true` because it's not used, we only analyse keys.

Wirone commented on src/Fixer/ClassNotation/FinalInternalClassFixer.php:
```diff
@@ -202,12 +260,93 @@ private function isClassCandidate(TokensAnalyzer $tokensAnalyzer, Tokens $tokens
+    {
+        $attributeCandidates = [];
+
+        while ($tokens[$index]->isGivenKind(CT::T_ATTRIBUTE_CLOSE)) {
```
Can we add test case like this?

```php
#[A]
/**
 * @B
 */
#[C]
class Foo {}
```

I really hope no-one is coding like this, but looking at the current implementation I think it would only catch `C` attribute and wouldn't reach `A`, so it would be really great if we handle it properly 🙂. Or such test case may prove me wrong, which would be good too 😉.

Wirone reviewed (changes requested):
@andythorne sorry for the delay, I had busy time (PHPers Summit conference), I couldn't take a look earlier. I'm leaving some concerns to address/verify, after that we'll probably be done 🙂.

PS. Please fix issues reported by Mess Detector too, so CI is green.

Wirone commented on src/Fixer/ClassNotation/FinalInternalClassFixer.php:
```diff
@@ -135,24 +153,48 @@ protected function createConfigurationDefinition(): FixerConfigurationResolverIn
 
         return new FixerConfigurationResolver([
-            (new FixerOptionBuilder('annotation_include', 'Class level annotations tags that must be set in order to fix the class (case insensitive).'))
+            new DeprecatedFixerOption(
```
Nice catch @paulbalandan! 👍

andythorne commented:
Thanks for the feedback again! The code really wasn't going to work nicely in cases of mixed attribute and annotation order, so i've refactored it along these lines:

- Loop tokens preceding class
  - skip `readonly` and comment tokens
  - If the token is an attribute, check it is included (true), exclude (false), or no decision is made (null)
  - If the token is a comment, check it is included (true), exclude (false), or no decision is made (null)
  - else, break and stop looking back

Wirone commented on src/Fixer/ClassNotation/FinalInternalClassFixer.php:
```diff
@@ -202,12 +278,95 @@ private function isClassCandidate(TokensAnalyzer $tokensAnalyzer, Tokens $tokens
+            return true;
+        }
+
+        foreach ($this->configuration['include'] as $tag => $true) {
+            if (isset($attributes[$tag])) {
+                return true;
+            }
+        }
+
+        return false;
```
```suggestion
        return count(array_intersect_key($this->configuration['include'], $attributes)) > 0;
```

Wirone commented on src/Fixer/ClassNotation/FinalInternalClassFixer.php:
```diff
@@ -177,23 +214,62 @@ private function isClassCandidate(TokensAnalyzer $tokensAnalyzer, Tokens $tokens
 
-            foreach ($this->configuration['annotation_exclude'] as $tagStart => $true) {
+            foreach ($this->configuration['exclude'] as $tagStart => $true) {
                 if (str_starts_with($tag, $tagStart)) {
```
I know this is not your code, but this is wrong and maybe you want fix it while working on this? Consider class:

```php
/**
 * @foo
 * @abcdef
 */
final class A{}
```

With config `['include' => ['@foo'], 'exclude' => ['@abc']]` it won't mark class as `final`, even if it should because there's configured `include`, while `exclude` _should_ not match.

andythorne replied:
Fixed it :)

Wirone commented on src/Fixer/ClassNotation/FinalInternalClassFixer.php:
```diff
@@ -177,23 +214,62 @@ private function isClassCandidate(TokensAnalyzer $tokensAnalyzer, Tokens $tokens
+            }
+        }
+
+        if (false !== array_search(false, $decisions, true)) {
+            return false;
+        }
+
+        if (false !== array_search(true, $decisions, true)) {
+            return true;
+        }
+
+        return !\count($decisions) && $this->configuration['consider_absent_docblock_as_internal_class'];
```
```suggestion
        if (in_array(false, $decisions, true)) {
            return false;
        }

        return in_array(true, $decisions, true)
            || ([] === $decisions && $this->configuration['consider_absent_docblock_as_internal_class']);
```

- [PHP Inspections (EA Extended)](https://plugins.jetbrains.com/plugin/7622-php-inspections-ea-extended-) plugin for PHPStorm marks it as warning and suggests changing `array_search()` to simple `in_array()` and I agree with it 😉
- I prefer explicit `[] === $foo` rather than `!count()`, and there are several `[] ===` already in the codebase

Wirone reviewed (changes requested):
All code suggestions were tested locally by running `FinalInternalClassFixerTest`. I did not want to commit it and push to your branch, but I believe these are worth applying 🙂.

I think it's a final round of review, after this improvements it'll be ready for merging (at least from my perspective).

andythorne commented on tests/Fixer/ClassNotation/FinalInternalClassFixerTest.php:
```diff
@@ -314,6 +314,17 @@ class B{}
+',
+                null,
+                [
+                    'include' => ['Hello'],
```
🙈 Fixed!

andythorne replied:
I've pushed the changes but it looks like [github is having issues](https://www.githubstatus.com/)

Wirone commented:
Thank you very much @andythorne 🍻. It took quite some time, but I think it was fruitful review process with great outcome 👍.
